MaterialLib/MPL/Properties/PermeabilityMohrCoulombFailureIndexModel.cpp

```diff
@@ -0,0 +1,153 @@
+/**
+ * \file
+ * \copyright
+ * Copyright (c) 2012-2020, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ * Created on June 4, 2020, 10:13 AM
+ */
+
+#include "PermeabilityMohrCoulombFailureIndexModel.h"
+
+#include <boost/math/constants/constants.hpp>
+#include <cmath>
+
+#include "BaseLib/Error.h"
+#include "MaterialLib/MPL/Medium.h"
+#include "MathLib/KelvinVector.h"
+#include "MathLib/MathTools.h"
+#include "ParameterLib/CoordinateSystem.h"
+#include "ParameterLib/Parameter.h"
+
+namespace MaterialPropertyLib
+{
+template <int DisplacementDim>
+PermeabilityMohrCoulombFailureIndexModel<DisplacementDim>::
+    PermeabilityMohrCoulombFailureIndexModel(
+        std::string name, ParameterLib::Parameter<double> const& k0,
+        double const kr, double const b, double const c, double const phi,
+        ParameterLib::CoordinateSystem const* const local_coordinate_system)
+    : k0_(k0),
+      kr_(kr),
+      b_(b),
+      c_(c),
+      phi_(boost::math::constants::pi<double>() * phi / 180.0),
+      local_coordinate_system_(local_coordinate_system)
+{
+    name_ = std::move(name);
+}
+
+template <int DisplacementDim>
+void PermeabilityMohrCoulombFailureIndexModel<DisplacementDim>::checkScale()
+    const
+{
+    if (!std::holds_alternative<Medium*>(scale_))
+    {
+        OGS_FATAL(
+            "The property 'PermeabilityMohrCoulombFailureIndexModel' is "
+            "implemented on the 'medium' scale only.");
+    }
+}
+
+template <int DisplacementDim>
+PropertyDataType
+PermeabilityMohrCoulombFailureIndexModel<DisplacementDim>::value(
+    VariableArray const& variable_array,
+    ParameterLib::SpatialPosition const& pos, double const t,
+    double const /*dt*/) const
+{
+    constexpr int symmetric_tensor_size =
+        MathLib::KelvinVector::KelvinVectorDimensions<DisplacementDim>::value;
+    using SymmetricTensor = Eigen::Matrix<double, symmetric_tensor_size, 1>;
+
+    auto const& stress_vector = std::get<SymmetricTensor>(
+        variable_array[static_cast<int>(Variable::stress)]);
+    auto const& stress_tensor =
+        MathLib::symmetricTensorToFullTensor(stress_vector);
+
+    Eigen::SelfAdjointEigenSolver<Eigen::Matrix<double, 3, 3>>
+        eigenvalue_solver(stress_tensor);
+
+    // Principle stress
+    auto const sigma = eigenvalue_solver.eigenvalues();
+
+    int const max_principle_stress_id = DisplacementDim - 1;
+    double const f = std::fabs(
+        0.5 * (sigma[0] - sigma[max_principle_stress_id]) /
+        (c_ * std::cos(phi_) +
+         0.5 * (sigma[0] + sigma[max_principle_stress_id]) * std::sin(phi_)));
```

Comment:
@nagelt  New changes:
1. still use std::fabs(f)
2. for 2D plane strain problem, use sigma[0] and sigma[1]. This took me a while to figure out  for 2D benchmark.

Please take a look.

Reply:
ok ... if I understand your changes correctly, there may be a few issues to consider:

The Eigen docu says that eigenvalues are returned in increasing order so that if you use 
sigma[2] - sigma[0] for example, your maximum shear stress should remain positive simply thanks to sorting by Eigen ... if you want to stay safe, I can see why you would want to use std::fabs(f) around the maximum shear stress, but not the entire expression ... because this seems to raise some issues:

If you are on the compressive side of the MohrCoulomb apex (c/tan(phi)), then your denominator is also positive and you don't need a fabs there either ... if you're on the tensile side of the apex, introducing a fabs will decrease the distance to the maximum shear stress again (and hence f), but intuitively I'd expect f to keep increasing (because it's more damaging) ... so you need another distance measure there I think ... in other words: if your denominator turns negative, things get worse ... but taking fabs() and making it positive artificially makes things better ... this is unwanted behaviour I think

By using sigma[0] and sigma[1] you seem to choose the middle principal stress. This can vary depending on loading and material parameters (particularly Poisson's ratio and dilatancy). So if you want to ensure that the slip happens "in-plane", I think you need to pass (sigma_xx, sigma_yy, sigma_xy) to the eigendecomposition, instead of (sigma_xx,sigma_yy,sigma_zz,sigma_xy) and then dropping the highest eigenvalue (sigma[2]) ... But in the solid material models we make no such "trick" but let the stresses and boundary conditions play it out ...

Reply:
Thanks. Yes, you are right that || is not needed, and that I wanted to ensure that the slip happens "in-plane". For a plane strain test with high in-plane pressure and with off-plane stress purely caused by the Poisson ratio, f with  sigma[0] and  sigma[1]  (occasionally in-plane, i.e. eigenvalues of 2x2 matrix by sigma_xx, sigma_yy, sigma_xy ) and f   with  sigma[0] and  sigma[2] are all larger than one. I my previous consideration,  f with  sigma[0] and  sigma[1] was selected for its better magnitude for the model. Yes, the "trick" seems not for general problems. I will figure out a decent way to choose f soon.

Reply:
@nagelt 
> so all that's missing now is the case of c*cos(a) = 0.5(sig_3 + sig_1) sin(a) where the denominator vanishes (apex) to avoid a division by zero ... and how you treat the point beyond

I got a idea for the apex point: assuming a cutting of the apex by 0.0001 (or even smaller) of c  * atan(phi).

Reply:
@wenqing 
so your denominator would be max( epsilon, (c + tan(phi))*cos(phi) ? 
this would correspond to a cutoff of f at f_max as the denominator goes to 0 (which would now let f go to infinity) ... so effectively this would mean a maximum permeability in the model (which can be practical) ... would do the trick for now, I guess. Did I understand your intention correctly?

another idea could be formulating the criterion as a difference instead of as a ratio  (no problems with division by zero or f going to infty, easier use of an Euclidean distance to the apex) but this would be a different model to the one in the paper, of course ... but conceptually similar ...

Reply:
@nagelt  The idea is when sigma_m=(sigma_0+sigma_0), set sigma_m = epsilon * sigma_m, then the dominator  is epsilon * c. I will also think about your idea of a difference.

Reply:
@nagelt 
 New changes for stress point at the apex
1. introduced a parameter of maximum_permeability.
2. cut-off at the apex with a small offset,
3. if f is too large, i.e. exp(bf) is infinite, returns maximum_permeability (changed the unit test for this as well).
4. if the k = max (k,  maximum_permeability)

This would guarantee that the permeability is in a reasonable magnitude at the extremist stress statue.

Reply:
thanks for all your changes, Wenqing. 1, 3, 4 are useful ... for 2 I left another comment, which should be a small fix but give the parameters some meaning ... then I'm ok with it!